b2/raw_simulator.py

```diff
@@ -229,18 +370,38 @@ def upload_file(
         url_match = re.match(r'https://upload.example.com/([^/]*)/([^/]*)', upload_url)
         if url_match is None:
             raise BadUploadUrl(upload_url)
+        if len(self.upload_errors) != 0:
+            raise self.upload_errors.pop(0)
         bucket_id, upload_id = url_match.groups()
-        return self._get_bucket(bucket_id).upload_file(
+        bucket = self._get_bucket_by_id(bucket_id)
+        return bucket.upload_file(
             upload_id, upload_auth_token, file_name, content_length, content_type, content_sha1,
             file_infos, data_stream
         )
 
+    def upload_part(
+        self, upload_url, upload_auth_token, part_number, content_length, sha1_sum, input_stream
+    ):
+        re.compile('https://upload.example.com/part/([^/]*)')
```

Comment:
`RE_UPLOAD_PART = re.compile('https://upload.example.com/part/([^/]*)')` in scope of the module and then: